sdk/identity/azure-identity/src/managed_identity_source.cpp

```diff
@@ -5,24 +5,62 @@
 
 #include <azure/core/internal/environment.hpp>
 
+#include <azure/core/internal/diagnostics/log.hpp>
+
 #include <fstream>
 #include <iterator>
 #include <stdexcept>
 #include <utility>
 
 using namespace Azure::Identity::_detail;
+
 using Azure::Core::_internal::Environment;
+using Azure::Core::Diagnostics::Logger;
+using Azure::Core::Diagnostics::_internal::Log;
+
+namespace {
+std::string const IdentityPrefix = "Identity: ";
+std::string const CredPrefix = "ManagedIdentityCredential";
+
+std::string WithSourceMessage(std::string const& credSource)
+{
+  return std::string(" with ") + credSource + " source";
+}
+
+void PrintEnvNotSetUpMessage(std::string const& credSource)
+{
+  auto const logLevel = Logger::Level::Verbose;
+  if (Log::ShouldWrite(logLevel))
+  {
+    Log::Write(
+        logLevel,
+        IdentityPrefix + CredPrefix + ": Environment is not set up for the credential to be created"
+            + WithSourceMessage(credSource) + '.');
+  }
+}
+} // namespace
 
 Azure::Core::Url ManagedIdentitySource::ParseEndpointUrl(
     std::string const& url,
-    char const* envVarName)
+    char const* envVarName,
+    std::string const credSource)
 {
   using Azure::Core::Url;
   using Azure::Core::Credentials::AuthenticationException;
 
   try
   {
-    return Url(url);
+    auto const endpointUrl = Url(url);
+
+    auto const logLevel = Logger::Level::Informational;
```

Comment:
Why do you need `Log::ShouldWrite` here? Doesn't `Log::Write` handle that case?

Or are you worried about the performance consequences of creating the log string? It seems like this might be an example of premature optimization - to my knowledge, we don't have a similar pattern in the rest of the codebasae, do we?

Reply:
Yes, in this case, this is to save that small amount of code on logging; `if(ShouldWrite()){Write()}` seem to be a pattern however, I've seen it in many places.

Reply:
#4421